source/h2_connection.c

```diff
@@ -623,6 +624,19 @@ static void s_on_channel_write_complete(
 
     CONNECTION_LOG(TRACE, connection, "Message finished writing to network. Rescheduling outgoing frame task");
 
+    while (!aws_linked_list_empty(&connection->thread_data.finish_encoding_streams_list)) {
+        struct aws_linked_list_node *node =
+            aws_linked_list_pop_front(&connection->thread_data.finish_encoding_streams_list);
+        struct aws_h2_stream *stream = AWS_CONTAINER_OF(node, struct aws_h2_stream, node);
+        AWS_ASSERT(stream->base.metrics.send_start_timestamp_ns != 0);
+        AWS_ASSERT(stream->base.metrics.send_end_timestamp_ns == 0);
+        aws_high_res_clock_get_ticks(&stream->base.metrics.send_end_timestamp_ns);
```

Comment:
it might be possible that we haven't set send_end_timestamp_ns yet, when we fire the completion callback

ughghhghgh I vote we just keep this simple, and forget about completion callbacks, and just take the time at our existing state-machine transitions, instead of adding complexity with more states

Reply:
We had another discussion later yesterday:

What about set the `send_end_timestamp_ns` from right before invoke complete callback if not set before.

So that, the `send_end_timestamp_ns` will be a bit off the real time at the corner case. We can document it out.

User will properly be interested in the time when the first byte is sent to the wire.